{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-----------------------\n",
    "# Required libs\n",
    "#-----------------------\n",
    "require(Seurat)\n",
    "require(sctransform)\n",
    "require(DropletUtils)\n",
    "require(PCAtools)\n",
    "require(biomaRt)\n",
    "require(stringr)\n",
    "\n",
    "\n",
    "runSeuratPipeline <- function(seuratdir, outdir = \"./\", norm = \"sctransform\",\n",
    "  mitoCutoff = 20,featuresLower = 100L,\n",
    "  featuresUpper = 10000L,countsLower = 1000L,\n",
    "  countsUpper = 100000L, dims = 1:30, verbose = T,\n",
    "  resolution = c(0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0),\n",
    "  samplename = \"seurat_pipeline_default\",\n",
    "  regressVars = c(), #c(\"percent.mito\", \"S.Score\", \"G2M.Score\"),\n",
    "  cellcycleGenes = NULL, species = \"mouse\", project = \"scRNAseq\",\n",
    "  meta = NULL, seed.use  = 10101){\n",
    "  \n",
    "    #-----------------------\n",
    "    # Run pipeline\n",
    "    #-----------------------\n",
    "    if(is.null(seuratdir)){\n",
    "        message(\"no input directory given. quitting\")\n",
    "        quit(status = 1)\n",
    "    } else {\n",
    "        counts.dir = file.path(seuratdir)\n",
    "    }\n",
    "    message(\"-------------------------\")\n",
    "    message(\"Seurat scRNAseq pipeline\")\n",
    "    message(\"-------------------------\")\n",
    "    message(paste(\"Processing:\", samplename))\n",
    "    message(paste(\"Input directory:\", seuratdir))\n",
    "    message(paste(\"Output directory:\", outdir))\n",
    "    message(paste(\"Species:\", species))\n",
    "    message(paste(\"Project:\", project))\n",
    "    message(paste(\"Mitochondrial % cutoff:\", mitoCutoff))\n",
    "    message(paste(\"# features cutoff:\", featuresLower, featuresUpper))\n",
    "    message(paste(\"# read counts cutoff:\", countsLower, countsUpper))\n",
    "    message(paste(\"Variables to regress during normalisation:\", toString(regressVars)))\n",
    "    message(paste(\"PCA dimensions to use:\", toString(dims)))\n",
    "    message(paste(\"Clustering resolutions to use:\", toString(resolution)))\n",
    "    message(paste(\"Metadata:\", meta))\n",
    "    message(paste(\"Random seed:\", seed.use))\n",
    "    \n",
    "    # setup output dirs\n",
    "    output.dir <- file.path(outdir)\n",
    "    dir.create(output.dir, showWarnings = F)\n",
    "\n",
    "    plot.dir <- file.path(outdir, \"plots\")\n",
    "    dir.create(plot.dir, showWarnings = F)\n",
    "\n",
    "    qc.dir <- file.path(outdir, \"qc\")\n",
    "    dir.create(qc.dir, showWarnings = F)\n",
    "\n",
    "    Seurat.obj.dir <- file.path(outdir, \"seurat_obj\")\n",
    "    dir.create(Seurat.obj.dir, showWarnings = F)\n",
    "\n",
    "    message(\"\")\n",
    "    message(\"Importing count data\")\n",
    "    \n",
    "    counts <- read.csv(paste0(dir, \"/counts.csv\"), header = FALSE)\n",
    "    metadata <- read.csv(paste0(dir, \"/metadata.csv\"), header = TRUE)\n",
    "    genes <- read.csv(paste0(dir, \"/genes.csv\"), header = FALSE)\n",
    "    cells <- read.csv(paste0(dir, \"/cells.csv\"), header = FALSE)\n",
    "    colnames(counts)<-cells$V1\n",
    "    rownames(counts)<-genes$V1\n",
    "    rownames(metadata)<-cells$V1\n",
    "    metadata<-metadata[c(\"Genotype\", \"Sample\", \"Timepoint\")]\n",
    "\n",
    "# Initialize the Seurat object with the raw (non-normalized data)\n",
    "\n",
    "    sample <- CreateSeuratObject(counts = counts, min.cells = 3, min.features = 200, project = project, meta.data = metadata)\n",
    "  Idents(sample)<-metadata$Sample\n",
    "    sample@meta.data$Sample = metadata$Sample\n",
    "    sample@meta.data$Genotype = metadata$Genotype\n",
    "    sample@meta.data$Timepoint = metadata$Timepoint\n",
    "    #-----------------------\n",
    "  # Sample QC\n",
    "  #-----------------------\n",
    "  message(\"Performing sample QC\")\n",
    "  mito.pattern <- \"^mt-\"\n",
    "  if(species == \"human\" | species == \"Human\" | species == \"Hs\" | species == \"hs\"){\n",
    "    mito.pattern <- \"^MT-\"\n",
    "  }\n",
    "\n",
    "  # calculate mito percentage\n",
    "  message(\"Calculating mito %\")\n",
    "  mito.genes <- grep(pattern = mito.pattern, x = rownames(x = sample@assays$RNA), value = TRUE)\n",
    "  percent.mito <- Matrix::colSums(sample@assays$RNA[mito.genes, ])/Matrix::colSums(sample@assays$RNA)*100\n",
    "  sample <- AddMetaData(object = sample, metadata = percent.mito, col.name = \"percent.mito\")\n",
    "    \n",
    "  # plot QC metrics\n",
    "  message(\"Plotting QC metrics - Pre-filter\")\n",
    "  pdf(file.path(qc.dir,paste(samplename, \"_basic_metrics_pre-filter.pdf\", sep = '')), width = 9, height = 7, useDingbats = F)\n",
    "  print(VlnPlot(object = sample, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3, pt.size = 0.01, log = T))\n",
    "  dev.off()\n",
    "\n",
    "  # subset based on mito count and feature cutoffs\n",
    "  message(\"Filtering low quality cells\")\n",
    "  message(\"Starting dimensions\")\n",
    "  print(dim(sample))\n",
    "  sample <- subset(sample, cells = colnames(sample)[which(sample$percent.mito < mitoCutoff)])\n",
    "  sample <- subset(sample, cells = colnames(sample)[which(sample$nCount_RNA > countsLower)])\n",
    "  sample <- subset(sample, cells = colnames(sample)[which(sample$nCount_RNA < countsUpper)])\n",
    "  sample <- subset(sample, cells = colnames(sample)[which(sample$nFeature_RNA > featuresLower)])\n",
    "  sample <- subset(sample, cells = colnames(sample)[which(sample$nFeature_RNA < featuresUpper)])\n",
    "\n",
    "  message(\"\")\n",
    "  message(\"Filtered dimensions\")\n",
    "  print(dim(sample))\n",
    "\n",
    "  # plot QC metrics\n",
    "  message(\"Plotting QC metrics - Post-filter\")\n",
    "  pdf(file.path(qc.dir,paste(samplename, \"_basic_metrics_post-filter.pdf\", sep = '')), width = 9, height = 7, useDingbats = F)\n",
    "  print(VlnPlot(object = sample, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3, pt.size = 0.01, log = T))\n",
    "  dev.off()\n",
    "    \n",
    "  # Feature Scatter plots\n",
    "  pdf(file.path(qc.dir,paste(samplename, \"_count_mito_QCscatter-postfilter.pdf\", sep = '')), width = 9, height = 7, useDingbats = F)\n",
    "  print(FeatureScatter(sample, feature1 = \"nCount_RNA\", feature2 = \"percent.mito\"))\n",
    "  dev.off()\n",
    "  pdf(file.path(qc.dir,paste(samplename, \"_count_feature_QCscatter-post-filter.pdf\", sep = '')), width = 9, height = 7, useDingbats = F)\n",
    "  print(FeatureScatter(sample, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\"))\n",
    "  dev.off()\n",
    "    \n",
    "     #-----------------------\n",
    "  # Normalisation #1\n",
    "  #-----------------------\n",
    "  message(paste0(\"Normalising counts using: \", norm))\n",
    "  # SCTransform workflow\n",
    "  message(\"Regressing following variables: percent.mito\")\n",
    "\n",
    "  if(norm == \"sctransform\") {\n",
    "    sample <- SCTransform(sample, vars.to.regress = \"percent.mito\", verbose = T, conserve.memory = F, return.only.var.genes = F, seed.use = seed.use)\n",
    "  }\n",
    "\n",
    "  # log normalisation workflow\n",
    "  if(norm == \"lognorm\") {\n",
    "    sample <- NormalizeData(sample, normalization.method = \"LogNormalize\",\n",
    "                              vars.to.regress = regressVars,\n",
    "                              verbose = F, conserve.memory = F)\n",
    "  }\n",
    "    \n",
    "     # log normalisation workflow\n",
    "  if(norm == \"RC\") {\n",
    "    sample <- NormalizeData(sample, normalization.method = \"RC\",\n",
    "                              vars.to.regress = regress.vars,\n",
    "                              verbose = F, conserve.memory = F)\n",
    "  }\n",
    "\n",
    "  #----------------------------------------------\n",
    "  # Seurat workflow without cell cycle regression\n",
    "  #----------------------------------------------\n",
    "  message(\"Seurat workflow without cell cycle regression\")\n",
    "\n",
    "  # run PCA\n",
    "  message(\"Running PCA\")\n",
    "  sample <- RunPCA(sample, npcs = 100, verbose = T, seed.use = seed.use)\n",
    "\n",
    "  # Run tSNE & UMAP\n",
    "  message(\"Running UMAP\")\n",
    "  sample <- RunUMAP(object = sample, verbose = T, seed.use = seed.use, dims = dims, reduction.name=paste(\"umap\",tail(dims,1), sep='_'))\n",
    "\n",
    "  message(\"Running TSNE\")\n",
    "  sample <- RunTSNE(object = sample, verbose = T, seed.use = seed.use, dims = dims, reduction.name=paste(\"tsne\",tail(dims,1), sep='_'))\n",
    "\n",
    "  message(\"Finding nearest neighbors\")\n",
    "  sample <- FindNeighbors(object = sample, verbose = T, dims = dims)\n",
    "\n",
    "  message(\"Running cluster sequence\")\n",
    "  # run a sequence of cluster resolutions\n",
    "  for (i in 1:length(resolution)){\n",
    "    sample <- FindClusters(object = sample, verbose = T, resolution = resolution[i], random.seed = seed.use)\n",
    "  }\n",
    "\n",
    "  message(\"Saving normalised dataset pre cell cycle regression\")\n",
    "  # save normalised and clustered dataset\n",
    "  saveRDS(sample, file = file.path(Seurat.obj.dir, paste(samplename, \"_normalised_no_cc_regression.rds\", sep = \"\")))\n",
    "\n",
    "  #----------------------------------------------\n",
    "  # Seurat workflow with cell cycle regression\n",
    "  #----------------------------------------------\n",
    "\n",
    "  #-----------------------\n",
    "  # Cell cycle scoring\n",
    "  #-----------------------\n",
    "\n",
    "  message(\"Performing Cell cycle scoring\")\n",
    "  # perform cell cycle analysis\n",
    "  s.genes <- Seurat::cc.genes.updated.2019$s.genes\n",
    "  g2m.genes <- Seurat::cc.genes.updated.2019$g2m.genes\n",
    "\n",
    "  if (species == \"mouse\"){\n",
    "    message(\"Species is mouse. Converting human to mouse gene symbols\")\n",
    "    convertHumanGeneList <- function(x){\n",
    "      human = useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\")\n",
    "      mouse = useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\")\n",
    "      genesV2 = getLDS(attributes = c(\"hgnc_symbol\"), filters = \"hgnc_symbol\",\n",
    "        values = x , mart = human, attributesL = c(\"mgi_symbol\"),\n",
    "        martL = mouse, uniqueRows=T)\n",
    "      humanx <- unique(genesV2[, 2])\n",
    "      return(humanx)\n",
    "    }\n",
    "    s.genes <- convertHumanGeneList(s.genes)\n",
    "    g2m.genes <- convertHumanGeneList(g2m.genes)\n",
    "    message(\"mouse cell cycle genes to use: S Phase\")\n",
    "    print(s.genes)\n",
    "    message(\"mouse cell cycle genes to use: G2M Phase\")\n",
    "    print(g2m.genes)\n",
    "  }\n",
    "\n",
    "  sample <- CellCycleScoring(sample, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE, search = F)\n",
    "\n",
    "  # add in cell cycle difference scoring\n",
    "  sample$CC.Difference <- sample$S.Score - sample$G2M.Score\n",
    "\n",
    "  # run PCA on cell cycle components\n",
    "  sample <- RunPCA(sample, features = c(s.genes, g2m.genes))\n",
    "\n",
    "  pdf(file.path(plot.dir,paste(samplename, \"_PCA_cell_cycle_anno.pdf\", sep = '')), width = 9, height = 7, useDingbats = F)\n",
    "  print(DimPlot(sample, reduction = \"pca\", group.by = 'Phase', label = T, label.size = 5))\n",
    "  dev.off()\n",
    "\n",
    "  #-----------------------\n",
    "  # Normalisation #2\n",
    "  #-----------------------\n",
    "  message(\"Normalisation #2\")\n",
    "  message(paste(\"Normalising counts using:\", norm))\n",
    "  # SCTransform workflow\n",
    "  message(\"Regressing following variables:\")\n",
    "\n",
    "  regressVars <- str_split(regressVars, ',')[[1]]\n",
    "  print(regressVars)\n",
    "\n",
    "  if(norm == \"sctransform\") {\n",
    "    sample <- SCTransform(sample, vars.to.regress = regressVars, verbose = T, conserve.memory = F, return.only.var.genes = F)\n",
    "  }\n",
    "\n",
    "  # log normalisation workflow\n",
    "  if(norm == \"lognorm\") {\n",
    "    sample <- NormalizeData(sample, normalization.method = \"LogNormalize\",\n",
    "                              vars.to.regress = regressVars,\n",
    "                              verbose = F, conserve.memory = F)\n",
    "  }\n",
    "\n",
    "  #--- Dimensional reduction & clustering\n",
    "  message(\"Running dimensional reduction algos\")\n",
    "\n",
    "  # always calculate 100 PCs\n",
    "  sample <- RunPCA(object = sample, verbose = T, npcs = 100, seed.use = seed.use)\n",
    "  message(\"Plotting...\")\n",
    "\n",
    "  # Elbow plot\n",
    "  percent.var <- Stdev(sample)\n",
    "  elbow.dim <- PCAtools::findElbowPoint(percent.var)\n",
    "  message(paste(\"Predicted elbow PC =\", elbow.dim))\n",
    "  pdf(file.path(plot.dir,paste(samplename,\"_elbow_plot.pdf\", sep = '')), width = 8, height = 6, useDingbats = F)\n",
    "  print(ElbowPlot(sample, ndims = 50) +\n",
    "    geom_vline(aes(xintercept = elbow.dim), color = \"red\") +\n",
    "    labs(title = paste(\"Elbow plot -\",samplename), subtitle = paste(\"Predicted elbow PC =\", elbow.dim)))\n",
    "  dev.off()\n",
    "\n",
    "  # Run tSNE & UMAP on 1 to elbow dimensions\n",
    "  message(\"Running UMAP\")\n",
    "  sample <- RunUMAP(object = sample, verbose = T, seed.use = seed.use, dims = 1:elbow.dim, reduction.name=\"umap_elbow\")\n",
    "  message(\"Running TSNE\")\n",
    "  sample <- RunTSNE(object = sample, verbose = T, seed.use = seed.use, dims = 1:elbow.dim, reduction.name=\"tsne_elbow\")\n",
    "\n",
    "  # find neighbors based on 1 to elbow dimensions\n",
    "  sample <- FindNeighbors(object = sample, verbose = T, dims = 1:elbow.dim)\n",
    "\n",
    "  # Run tSNE & UMAP on 1 to dims\n",
    "  message(\"Running UMAP\")\n",
    "  sample <- RunUMAP(object = sample, verbose = T, seed.use = seed.use, dims = dims, reduction.name=paste(\"umap\",tail(dims,1), sep='_'))\n",
    "  message(\"Running TSNE\")\n",
    "  sample <- RunTSNE(object = sample, verbose = T, seed.use = seed.use, dims = dims, reduction.name=paste(\"tsne\",tail(dims,1), sep='_'))\n",
    "\n",
    "  # run a sequence of cluster resolutions for elbow dim\n",
    "  for (i in 1:length(resolution)){\n",
    "    sample <- FindClusters(object = sample, verbose = T, resolution = resolution[i], random.seed = seed.use)\n",
    "  }\n",
    "\n",
    "  # save normalised and clustered dataset\n",
    "  message(\"Saving normalised dataset post cell cycle regression\")\n",
    "  saveRDS(sample, file = file.path(Seurat.obj.dir, paste(samplename, \"_cc_regressed_normalised.rds\", sep = \"\")))\n",
    "\n",
    "  message(\"performing final plots on normalised dataset\")\n",
    "\n",
    "  # PCA plot reduction on clusters\n",
    "  p <- DimPlot(object = sample, reduction = \"pca\", label = T, label.size = 5, dims = c(1,2)) +\n",
    "    NoLegend() +\n",
    "    ggtitle(paste('PCA -', samplename))\n",
    "\n",
    "  pdf(file.path(plot.dir,paste(samplename, \"_PCA.pdf\", sep = '')), width = 9, height = 7)\n",
    "  print(AugmentPlot(p, dpi = 600, width = 8, height = 6))\n",
    "  dev.off()\n",
    "\n",
    "  # UMAP reduction on clusters - elbow\n",
    "  p <- DimPlot(object = sample, reduction = \"umap_elbow\", label = T, label.size = 5) +\n",
    "    NoLegend() +\n",
    "    ggtitle(paste('UMAP -', samplename))\n",
    "\n",
    "  pdf(file.path(plot.dir,paste(samplename, \"_UMAP_elbow.pdf\", sep = '')), width = 9, height = 7)\n",
    "  print(AugmentPlot(p, dpi = 600, width = 8, height = 6))\n",
    "  dev.off()\n",
    "\n",
    "  # t-SNE reduction on clusters - elbow\n",
    "  p <- DimPlot(object = sample, reduction = \"tsne_elbow\", label = T, label.size = 5) +\n",
    "    NoLegend() +\n",
    "    ggtitle(paste('t-SNE -', samplename))\n",
    "\n",
    "  pdf(file.path(plot.dir,paste(samplename, \"_t-SNE_elbow.pdf\", sep = '')), width = 9, height = 7)\n",
    "  print(AugmentPlot(p, dpi = 600, width = 8, height = 6))\n",
    "  dev.off()\n",
    "\n",
    "\n",
    "     Idents(sample)<-metadata$Sample\n",
    "    \n",
    "    \n",
    "      # PCA plot reduction on clusters\n",
    "  p <- DimPlot(object = sample, reduction = \"pca\", label = T,, label.size = 5, dims = c(1,2)) +\n",
    "\n",
    "    ggtitle(paste('PCA -', samplename))\n",
    "\n",
    "  pdf(file.path(plot.dir,paste(samplename, \"_PCAsample.pdf\", sep = '')), width = 9, height = 7)\n",
    "  print(AugmentPlot(p, dpi = 600, width = 8, height = 6))\n",
    "  dev.off()\n",
    "\n",
    "  # UMAP reduction on clusters - elbow\n",
    "  p <- DimPlot(object = sample, reduction = \"umap_elbow\", label = T, label.size = 5) +\n",
    "\n",
    "    ggtitle(paste('UMAP -', samplename))\n",
    "\n",
    "  pdf(file.path(plot.dir,paste(samplename, \"_UMAP_elbowsample.pdf\", sep = '')), width =9, height = 7)\n",
    "  print(AugmentPlot(p, dpi = 600, width = 8, height = 6))\n",
    "  dev.off()\n",
    "\n",
    "  # t-SNE reduction on clusters - elbow\n",
    "  p <- DimPlot(object = sample, reduction = \"tsne_elbow\", label = T, label.size = 5) +\n",
    " \n",
    "    ggtitle(paste('t-SNE -', samplename))\n",
    "\n",
    "  pdf(file.path(plot.dir,paste(samplename, \"_t-SNE_elbowsample.pdf\", sep = '')), width = 9, height = 7)\n",
    "  print(AugmentPlot(p, dpi = 600, width = 8, height = 6))\n",
    "  dev.off()\n",
    "\n",
    " \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "  message(paste(\"Finished\", samplename, \"analysis. Enjoy your data!\"))\n",
    "\n",
    "  sessionInfo()\n",
    "}\n",
    "\n",
    "\n",
    "dir <- \"SergiSeurat\"\n",
    "\n",
    "\n",
    "runSeuratPipeline(dir, \n",
    "                  outdir = \"./output_seurat\"\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
