{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n",
      "┌ Warning: RCall.jl: estimating size factors\n",
      "│ estimating dispersions\n",
      "│ gene-wise dispersion estimates\n",
      "│ mean-dispersion relationship\n",
      "│ final dispersion estimates\n",
      "│ fitting model and testing\n",
      "└ @ RCall /Users/irr15/.julia/packages/RCall/iMDW2/src/io.jl:160\n"
     ]
    }
   ],
   "source": [
    "using RCall\n",
    "R\"\"\"\n",
    "###################################\n",
    "# DESeq2 GROseq enhancer analysis #\n",
    "###################################\n",
    "\n",
    "# Back in December 2015 Gopu did an analysis on the \n",
    "# Sergi GROseq macrophage reads. However he only compared\n",
    "# WT to Rad21KO macrophages. In order to make a figure stating that changes in gene expression upon LPS occur through changes in transcription, I needed comparisons accross timepoints.\n",
    "# As I had Gopu original code, I had just to track down the files in the ax4, which were in:\n",
    "\n",
    "#.     - /csc/analysis/Merkenshlager/SergiCuartero/Groseq_Dec2015/Merged\n",
    "\n",
    "# And then I adapted the code to make the comparisons I needed\n",
    "\n",
    "library(dplyr)\n",
    "library(DESeq2);\n",
    "library(IRanges)\n",
    "library(GenomicRanges)\n",
    "\n",
    "\n",
    "load(\"../Databases/Cuartero2018/GROseq/OstuneEnhModified_Counts_Dec2015.RData\")\n",
    "deseq <- function(sample1, sample2){\n",
    "    GROEnhCounts <- EnhancerSummaryCounts\n",
    "colnames(GROEnhCounts) <- gsub(\".merged.dupMarkd.bam\", \"\", colnames(GROEnhCounts))\n",
    "    group <- factor(substr(colnames(GROEnhCounts), 1, 3))  # for all samples\n",
    "    group <- relevel(group, sample2)\n",
    "\n",
    "    condition <- data.frame(condition = group, row.names = colnames(GROEnhCounts))\n",
    "    dds <- DESeqDataSetFromMatrix(countData = GROEnhCounts, colData = condition, \n",
    "    design = ~condition)\n",
    "\n",
    "    GROWTdds <- DESeq(dds)\n",
    "    subdds_results <- results(GROWTdds, contrast=c(\"condition\",sample1,sample2),alpha=0.1)\n",
    "    subdds_results<- cbind(as.data.frame(OstuniEnhGR_modified), as.data.frame(subdds_results))\n",
    "    write.table(subdds_results, paste0(\"../Databases/Cuartero2018/GROseq/GROseq_\",sample1, \"__vs__\", sample2, \"__DESeq2_OtsuniEnhancers.csv\"),sep=\",\",quote=F,row.names=F)\n",
    "    return(subdds_results)\n",
    "}\n",
    "\n",
    "deseq(\"WT1\",\"WT0\");\n",
    "deseq(\"WT6\",\"WT0\");\n",
    "deseq(\"FL0\",\"WT0\");\n",
    "deseq(\"FL1\",\"WT1\");\n",
    "deseq(\"FL6\",\"WT6\");\n",
    "deseq(\"WT6\",\"WT1\");\n",
    "deseq(\"FL1\",\"FL0\");\n",
    "deseq(\"FL6\",\"FL0\");\n",
    "deseq(\"FL6\",\"FL1\");\n",
    "\"\"\";"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m      Status\u001b[22m\u001b[39m `~/.julia/environments/v1.6/Project.toml`\n",
      " \u001b[90m [7e6ae17a] \u001b[39mBioSequences v2.0.5\n",
      " \u001b[90m [336ed68f] \u001b[39mCSV v0.9.11\n",
      " \u001b[90m [8f4d0f93] \u001b[39mConda v1.5.2\n",
      " \u001b[90m [a93c6f00] \u001b[39mDataFrames v1.2.2\n",
      " \u001b[90m [31c24e10] \u001b[39mDistributions v0.25.28\n",
      " \u001b[90m [c2308a5c] \u001b[39mFASTX v1.2.0\n",
      " \u001b[90m [09f84164] \u001b[39mHypothesisTests v0.10.6\n",
      " \u001b[90m [7073ff75] \u001b[39mIJulia v1.23.2\n",
      " \u001b[90m [916415d5] \u001b[39mImages v0.24.1\n",
      " \u001b[90m [f8716d33] \u001b[39mMultipleTesting v0.5.0\n",
      " \u001b[90m [92933f4c] \u001b[39mProgressMeter v1.7.1\n",
      " \u001b[90m [d330b81b] \u001b[39mPyPlot v2.10.0\n",
      " \u001b[90m [6f49c342] \u001b[39mRCall v0.13.12\n",
      " \u001b[90m [d2ef9438] \u001b[39mSeaborn v1.0.0\n",
      " \u001b[90m [2913bbd2] \u001b[39mStatsBase v0.33.13\n"
     ]
    }
   ],
   "source": [
    "import Pkg; Pkg.status()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.3",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
