{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scRNAseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ENV[\"Code\"] = \"../../Code\"\n",
    "for folder in readdir(ENV[\"Code\"]); push!(LOAD_PATH, normpath(ENV[\"Code\"], folder));end\n",
    "\n",
    "using Seaborn, PyPlot, PyCall, DataFrames\n",
    "using NoLongerProblems_Pandas, Pandas\n",
    "using MultipleTesting, HypothesisTests \n",
    "using SingleCellExperiment\n",
    "using CSV, ProgressMeter,PrettyPlotting, ScikitLearn\n",
    "@sk_import linear_model: LogisticRegression\n",
    "\n",
    "include(\"../Databases/Cuartero2018.jl\")\n",
    "include(\"../Databases/Bhatt2012.jl\")\n",
    "include(\"../Databases/MF_SingleCell.jl\")\n",
    "\n",
    "include(\"Load_scRNAseqData.jl\")\n",
    "\n",
    "sce = SingleCellExperiment.fit_mu_std_alpha(sce, splitdataby = :Sample, assay = \"CPM\")\n",
    "sce = SingleCellExperiment.fit_single_cell_logistic_regression(sce, assay = \"CPM\")\n",
    "bhattgenes = Bhatt2012.inducible_genes_figure3()[!,:GeneSymbol];\n",
    "sce = SingleCellExperiment.Shalek2014_module_score(collect(bhattgenes), sce,fitparameter = \"mu\", modulescore_name = :BhattGenesScore, untreated_pattern = \"UT\",comparedtothissample = \"WT\", assay = \"CPM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>GeneID</th><th>TotalCounts</th><th>TotalCells</th><th>FractionCells</th><th>WT_UT__mu</th><th>WT_UT__std</th><th>WT_UT__alpha</th></tr><tr><th></th><th>String31</th><th>Float64</th><th>Int64</th><th>Float64</th><th>Any</th><th>Any</th><th>Any</th></tr></thead><tbody><p>497 rows × 40 columns (omitted printing of 33 columns)</p><tr><th>1</th><td>Ccnd2</td><td>279.0</td><td>164</td><td>0.00889902</td><td>7.17715</td><td>0.604247</td><td>0.0376569</td></tr><tr><th>2</th><td>Slfn4</td><td>117.0</td><td>88</td><td>0.00477508</td><td>7.40551</td><td>0.651266</td><td>0.0334728</td></tr><tr><th>3</th><td>Lhx2</td><td>80.0</td><td>63</td><td>0.00341853</td><td>7.35961</td><td>0.312568</td><td>0.0083682</td></tr><tr><th>4</th><td>Mx1</td><td>552.0</td><td>260</td><td>0.0141082</td><td>7.49252</td><td>0.680389</td><td>0.0753138</td></tr><tr><th>5</th><td>Fap</td><td>75.0</td><td>64</td><td>0.00347279</td><td>7.51507</td><td>0.522559</td><td>0.0292887</td></tr><tr><th>6</th><td>Itga5</td><td>1022.0</td><td>513</td><td>0.0278366</td><td>7.34853</td><td>0.426283</td><td>0.217573</td></tr><tr><th>7</th><td>Icosl</td><td>149.0</td><td>84</td><td>0.00455803</td><td>7.31429</td><td>0.18186</td><td>0.0292887</td></tr><tr><th>8</th><td>Il12rb1</td><td>30.0</td><td>26</td><td>0.00141082</td><td>7.40171</td><td>0.326519</td><td>0.0209205</td></tr><tr><th>9</th><td>Mmp14</td><td>392.0</td><td>182</td><td>0.00987574</td><td>6.88524</td><td>1.12759</td><td>0.0083682</td></tr><tr><th>10</th><td>Ccl3</td><td>35750.0</td><td>1215</td><td>0.0659287</td><td>8.14453</td><td>0.700594</td><td>0.707113</td></tr><tr><th>11</th><td>Mxd1</td><td>628.0</td><td>301</td><td>0.016333</td><td>7.39612</td><td>0.511915</td><td>0.0794979</td></tr><tr><th>12</th><td>Gramd1a</td><td>604.0</td><td>347</td><td>0.018829</td><td>7.47333</td><td>0.51896</td><td>0.108787</td></tr><tr><th>13</th><td>Acp5</td><td>102.0</td><td>57</td><td>0.00309295</td><td>7.77631</td><td>0.364022</td><td>0.0334728</td></tr><tr><th>14</th><td>Col18a1</td><td>179.0</td><td>127</td><td>0.00689131</td><td>7.33776</td><td>0.366235</td><td>0.121339</td></tr><tr><th>15</th><td>Ell2</td><td>1038.0</td><td>471</td><td>0.0255575</td><td>7.5229</td><td>0.505108</td><td>0.171548</td></tr><tr><th>16</th><td>Ifrd1</td><td>980.0</td><td>455</td><td>0.0246893</td><td>7.52617</td><td>0.580153</td><td>0.284519</td></tr><tr><th>17</th><td>Daxx</td><td>706.0</td><td>362</td><td>0.019643</td><td>7.51196</td><td>0.493922</td><td>0.230126</td></tr><tr><th>18</th><td>Axl</td><td>339.0</td><td>198</td><td>0.0107439</td><td>7.48847</td><td>0.365206</td><td>0.0920502</td></tr><tr><th>19</th><td>Lcp2</td><td>1221.0</td><td>569</td><td>0.0308753</td><td>7.43434</td><td>0.556433</td><td>0.213389</td></tr><tr><th>20</th><td>Ggct</td><td>53.0</td><td>43</td><td>0.00233328</td><td>6.95848</td><td>0.518217</td><td>0.0209205</td></tr><tr><th>21</th><td>Relb</td><td>674.0</td><td>374</td><td>0.0202941</td><td>7.49799</td><td>0.544747</td><td>0.16318</td></tr><tr><th>22</th><td>Ebi3</td><td>3433.0</td><td>889</td><td>0.0482392</td><td>7.68661</td><td>0.56119</td><td>0.405858</td></tr><tr><th>23</th><td>Ier3</td><td>1782.0</td><td>565</td><td>0.0306582</td><td>7.35999</td><td>0.58406</td><td>0.083682</td></tr><tr><th>24</th><td>Cp</td><td>430.0</td><td>243</td><td>0.0131857</td><td>7.46028</td><td>0.236472</td><td>0.0209205</td></tr><tr><th>25</th><td>Stat5a</td><td>220.0</td><td>145</td><td>0.00786803</td><td>7.37138</td><td>0.677247</td><td>0.0711297</td></tr><tr><th>26</th><td>Il12b</td><td>831.0</td><td>252</td><td>0.0136741</td><td>7.35696</td><td>0.340193</td><td>0.0209205</td></tr><tr><th>27</th><td>Ccl24</td><td>82.0</td><td>33</td><td>0.00179066</td><td>0.0</td><td>0.0</td><td>0.0</td></tr><tr><th>28</th><td>Sh2b2</td><td>126.0</td><td>93</td><td>0.00504639</td><td>7.06713</td><td>0.570288</td><td>0.0292887</td></tr><tr><th>29</th><td>Cd44</td><td>2054.0</td><td>742</td><td>0.0402626</td><td>7.80203</td><td>0.544473</td><td>0.556485</td></tr><tr><th>30</th><td>Rcn1</td><td>71.0</td><td>58</td><td>0.00314721</td><td>7.57592</td><td>0.117874</td><td>0.0125523</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccccccc}\n",
       "\t& GeneID & TotalCounts & TotalCells & FractionCells & WT\\_UT\\_\\_mu & WT\\_UT\\_\\_std & WT\\_UT\\_\\_alpha & \\\\\n",
       "\t\\hline\n",
       "\t& String31 & Float64 & Int64 & Float64 & Any & Any & Any & \\\\\n",
       "\t\\hline\n",
       "\t1 & Ccnd2 & 279.0 & 164 & 0.00889902 & 7.17715 & 0.604247 & 0.0376569 & $\\dots$ \\\\\n",
       "\t2 & Slfn4 & 117.0 & 88 & 0.00477508 & 7.40551 & 0.651266 & 0.0334728 & $\\dots$ \\\\\n",
       "\t3 & Lhx2 & 80.0 & 63 & 0.00341853 & 7.35961 & 0.312568 & 0.0083682 & $\\dots$ \\\\\n",
       "\t4 & Mx1 & 552.0 & 260 & 0.0141082 & 7.49252 & 0.680389 & 0.0753138 & $\\dots$ \\\\\n",
       "\t5 & Fap & 75.0 & 64 & 0.00347279 & 7.51507 & 0.522559 & 0.0292887 & $\\dots$ \\\\\n",
       "\t6 & Itga5 & 1022.0 & 513 & 0.0278366 & 7.34853 & 0.426283 & 0.217573 & $\\dots$ \\\\\n",
       "\t7 & Icosl & 149.0 & 84 & 0.00455803 & 7.31429 & 0.18186 & 0.0292887 & $\\dots$ \\\\\n",
       "\t8 & Il12rb1 & 30.0 & 26 & 0.00141082 & 7.40171 & 0.326519 & 0.0209205 & $\\dots$ \\\\\n",
       "\t9 & Mmp14 & 392.0 & 182 & 0.00987574 & 6.88524 & 1.12759 & 0.0083682 & $\\dots$ \\\\\n",
       "\t10 & Ccl3 & 35750.0 & 1215 & 0.0659287 & 8.14453 & 0.700594 & 0.707113 & $\\dots$ \\\\\n",
       "\t11 & Mxd1 & 628.0 & 301 & 0.016333 & 7.39612 & 0.511915 & 0.0794979 & $\\dots$ \\\\\n",
       "\t12 & Gramd1a & 604.0 & 347 & 0.018829 & 7.47333 & 0.51896 & 0.108787 & $\\dots$ \\\\\n",
       "\t13 & Acp5 & 102.0 & 57 & 0.00309295 & 7.77631 & 0.364022 & 0.0334728 & $\\dots$ \\\\\n",
       "\t14 & Col18a1 & 179.0 & 127 & 0.00689131 & 7.33776 & 0.366235 & 0.121339 & $\\dots$ \\\\\n",
       "\t15 & Ell2 & 1038.0 & 471 & 0.0255575 & 7.5229 & 0.505108 & 0.171548 & $\\dots$ \\\\\n",
       "\t16 & Ifrd1 & 980.0 & 455 & 0.0246893 & 7.52617 & 0.580153 & 0.284519 & $\\dots$ \\\\\n",
       "\t17 & Daxx & 706.0 & 362 & 0.019643 & 7.51196 & 0.493922 & 0.230126 & $\\dots$ \\\\\n",
       "\t18 & Axl & 339.0 & 198 & 0.0107439 & 7.48847 & 0.365206 & 0.0920502 & $\\dots$ \\\\\n",
       "\t19 & Lcp2 & 1221.0 & 569 & 0.0308753 & 7.43434 & 0.556433 & 0.213389 & $\\dots$ \\\\\n",
       "\t20 & Ggct & 53.0 & 43 & 0.00233328 & 6.95848 & 0.518217 & 0.0209205 & $\\dots$ \\\\\n",
       "\t21 & Relb & 674.0 & 374 & 0.0202941 & 7.49799 & 0.544747 & 0.16318 & $\\dots$ \\\\\n",
       "\t22 & Ebi3 & 3433.0 & 889 & 0.0482392 & 7.68661 & 0.56119 & 0.405858 & $\\dots$ \\\\\n",
       "\t23 & Ier3 & 1782.0 & 565 & 0.0306582 & 7.35999 & 0.58406 & 0.083682 & $\\dots$ \\\\\n",
       "\t24 & Cp & 430.0 & 243 & 0.0131857 & 7.46028 & 0.236472 & 0.0209205 & $\\dots$ \\\\\n",
       "\t25 & Stat5a & 220.0 & 145 & 0.00786803 & 7.37138 & 0.677247 & 0.0711297 & $\\dots$ \\\\\n",
       "\t26 & Il12b & 831.0 & 252 & 0.0136741 & 7.35696 & 0.340193 & 0.0209205 & $\\dots$ \\\\\n",
       "\t27 & Ccl24 & 82.0 & 33 & 0.00179066 & 0.0 & 0.0 & 0.0 & $\\dots$ \\\\\n",
       "\t28 & Sh2b2 & 126.0 & 93 & 0.00504639 & 7.06713 & 0.570288 & 0.0292887 & $\\dots$ \\\\\n",
       "\t29 & Cd44 & 2054.0 & 742 & 0.0402626 & 7.80203 & 0.544473 & 0.556485 & $\\dots$ \\\\\n",
       "\t30 & Rcn1 & 71.0 & 58 & 0.00314721 & 7.57592 & 0.117874 & 0.0125523 & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1m497×40 DataFrame\u001b[0m\n",
       "\u001b[1m Row \u001b[0m│\u001b[1m GeneID        \u001b[0m\u001b[1m TotalCounts \u001b[0m\u001b[1m TotalCells \u001b[0m\u001b[1m FractionCells \u001b[0m\u001b[1m WT_UT__mu \u001b[0m\u001b[1m WT_UT\u001b[0m ⋯\n",
       "\u001b[1m     \u001b[0m│\u001b[90m String31      \u001b[0m\u001b[90m Float64     \u001b[0m\u001b[90m Int64      \u001b[0m\u001b[90m Float64       \u001b[0m\u001b[90m Any       \u001b[0m\u001b[90m Any  \u001b[0m ⋯\n",
       "─────┼──────────────────────────────────────────────────────────────────────────\n",
       "   1 │ Ccnd2                279.0         164     0.00889902  7.17715    0.604 ⋯\n",
       "   2 │ Slfn4                117.0          88     0.00477508  7.40551    0.651\n",
       "   3 │ Lhx2                  80.0          63     0.00341853  7.35961    0.312\n",
       "   4 │ Mx1                  552.0         260     0.0141082   7.49252    0.680\n",
       "   5 │ Fap                   75.0          64     0.00347279  7.51507    0.522 ⋯\n",
       "   6 │ Itga5               1022.0         513     0.0278366   7.34853    0.426\n",
       "   7 │ Icosl                149.0          84     0.00455803  7.31429    0.181\n",
       "   8 │ Il12rb1               30.0          26     0.00141082  7.40171    0.326\n",
       "   9 │ Mmp14                392.0         182     0.00987574  6.88524    1.127 ⋯\n",
       "  10 │ Ccl3               35750.0        1215     0.0659287   8.14453    0.700\n",
       "  11 │ Mxd1                 628.0         301     0.016333    7.39612    0.511\n",
       "  ⋮  │       ⋮             ⋮           ⋮             ⋮            ⋮          ⋮ ⋱\n",
       " 488 │ Gm12250              221.0         169     0.00917033  7.47157    0.517\n",
       " 489 │ A530032D15Rik        179.0         125     0.00678279  7.59983    0.434 ⋯\n",
       " 490 │ Bcl2a1b             2166.0         603     0.0327202   7.23072    0.692\n",
       " 491 │ Cfb                  809.0         290     0.0157361   7.39981    0.158\n",
       " 492 │ Mndal               4030.0         945     0.0512779   7.88463    0.537\n",
       " 493 │ Hspa1b               186.0          81     0.00439525  8.21921    NaN   ⋯\n",
       " 494 │ Gm6904                75.0          65     0.00352705  7.1719     0.538\n",
       " 495 │ Gbp11                 90.0          63     0.00341853  7.90042    0.349\n",
       " 496 │ Kctd12               777.0         369     0.0200228   7.4629     0.522\n",
       " 497 │ Gbp5                 577.0         271     0.0147051   7.79389    0.608 ⋯\n",
       "\u001b[36m                                                 35 columns and 476 rows omitted\u001b[0m"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sce = SingleCellExperiment.fit_mu_std_alpha(sce, splitdataby = :Sample, assay = \"lnCPMplus1\")\n",
    "\n",
    "bhattgenes= DataFrames.DataFrame(\n",
    "    \"GeneID\" => Bhatt2012.inducible_genes_figure3()[!,:GeneSymbol], \n",
    "    \"Class\"=>Bhatt2012.inducible_genes_figure3()[!,:Class]\n",
    "    );\n",
    "sceBhatt = innerjoin(sce.rowData, bhattgenes,on = :GeneID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ln CPM plus 1 in expressing cells\n",
      "WT UT vs Rad21KO UT\n",
      "Approximate Wilcoxon signed rank test\n",
      "-------------------------------------\n",
      "Population details:\n",
      "    parameter of interest:   Location parameter (pseudomedian)\n",
      "    value under h_0:         0\n",
      "    point estimate:          0.170879\n",
      "    95% confidence interval: (0.1542, 0.1995)\n",
      "\n",
      "Test summary:\n",
      "    outcome with 95% confidence: reject h_0\n",
      "    two-sided p-value:           <1e-37\n",
      "\n",
      "Details:\n",
      "    number of observations:      497\n",
      "    Wilcoxon rank-sum statistic: 97837.0\n",
      "    rank sums:                   [97837.0, 18566.0]\n",
      "    adjustment for ties:         0.0\n",
      "    normal approximation (μ, σ): (39635.5, 3059.53)\n",
      "\n",
      "Fraction Expressing cells\n",
      "WT UT vs Rad21KO UT\n",
      "Approximate Wilcoxon signed rank test\n",
      "-------------------------------------\n",
      "Population details:\n",
      "    parameter of interest:   Location parameter (pseudomedian)\n",
      "    value under h_0:         0\n",
      "    point estimate:          0.0155742\n",
      "    95% confidence interval: (0.0196, 0.0297)\n",
      "\n",
      "Test summary:\n",
      "    outcome with 95% confidence: reject h_0\n",
      "    two-sided p-value:           <1e-27\n",
      "\n",
      "Details:\n",
      "    number of observations:      497\n",
      "    Wilcoxon rank-sum statistic: 91780.0\n",
      "    rank sums:                   [91780.0, 24623.0]\n",
      "    adjustment for ties:         1002.0\n",
      "    normal approximation (μ, σ): (33578.5, 3059.53)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "5.046706485254621e-28"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "println(\"ln CPM plus 1 in expressing cells\")\n",
    "println(\"WT UT vs Rad21KO UT\")\n",
    "\n",
    "t = HypothesisTests.SignedRankTest([ii for ii in sceBhatt[!,\"WT_UT__mu\"]], [ii for ii in sceBhatt[!,\"RAD21_UT__mu\"]])\n",
    "println(t)\n",
    "p1 = pvalue(t)\n",
    "\n",
    "\n",
    "println(\"Fraction Expressing cells\")\n",
    "println(\"WT UT vs Rad21KO UT\")\n",
    "\n",
    "t = HypothesisTests.SignedRankTest([ii for ii in sceBhatt[!,\"WT_UT__alpha\"]], [ii for ii in sceBhatt[!,\"RAD21_UT__alpha\"]])\n",
    "println(t)\n",
    "p1a = pvalue(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ln CPM plus 1 in expressing cells\n",
      "WT 2H vs Rad21KO 2H\n",
      "Approximate Wilcoxon signed rank test\n",
      "-------------------------------------\n",
      "Population details:\n",
      "    parameter of interest:   Location parameter (pseudomedian)\n",
      "    value under h_0:         0\n",
      "    point estimate:          0.0330481\n",
      "    95% confidence interval: (-0.0054, 0.0544)\n",
      "\n",
      "Test summary:\n",
      "    outcome with 95% confidence: fail to reject h_0\n",
      "    two-sided p-value:           0.1074\n",
      "\n",
      "Details:\n",
      "    number of observations:      497\n",
      "    Wilcoxon rank-sum statistic: 66770.0\n",
      "    rank sums:                   [66770.0, 56486.0]\n",
      "    adjustment for ties:         0.0\n",
      "    normal approximation (μ, σ): (5142.0, 3193.66)\n",
      "\n",
      "Fraction Expressing cells\n",
      "WT 2H vs Rad21KO 2H\n",
      "Approximate Wilcoxon signed rank test\n",
      "-------------------------------------\n",
      "Population details:\n",
      "    parameter of interest:   Location parameter (pseudomedian)\n",
      "    value under h_0:         0\n",
      "    point estimate:          0.0376344\n",
      "    95% confidence interval: (0.043, 0.0591)\n",
      "\n",
      "Test summary:\n",
      "    outcome with 95% confidence: reject h_0\n",
      "    two-sided p-value:           <1e-38\n",
      "\n",
      "Details:\n",
      "    number of observations:      497\n",
      "    Wilcoxon rank-sum statistic: 99099.5\n",
      "    rank sums:                   [99099.5, 18270.5]\n",
      "    adjustment for ties:         29904.0\n",
      "    normal approximation (μ, σ): (40414.5, 3078.47)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2.2779482647126536e-39"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "println(\"ln CPM plus 1 in expressing cells\")\n",
    "println(\"WT 2H vs Rad21KO 2H\")\n",
    "\n",
    "t = HypothesisTests.SignedRankTest([ii for ii in sceBhatt[!,\"WT_2H__mu\"]], [ii for ii in sceBhatt[!,\"RAD21_2H__mu\"]])\n",
    "println(t)\n",
    "p2 = pvalue(t)\n",
    "\n",
    "println(\"Fraction Expressing cells\")\n",
    "println(\"WT 2H vs Rad21KO 2H\")\n",
    "\n",
    "t = HypothesisTests.SignedRankTest([ii for ii in sceBhatt[!,\"WT_2H__alpha\"]], [ii for ii in sceBhatt[!,\"RAD21_2H__alpha\"]])\n",
    "println(t)\n",
    "p2a = pvalue(t)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ln CPM plus 1 in expressing cells\n",
      "WT 8H vs Rad21KO 8H\n",
      "Approximate Wilcoxon signed rank test\n",
      "-------------------------------------\n",
      "Population details:\n",
      "    parameter of interest:   Location parameter (pseudomedian)\n",
      "    value under h_0:         0\n",
      "    point estimate:          0.34752\n",
      "    95% confidence interval: (0.3302, 0.384)\n",
      "\n",
      "Test summary:\n",
      "    outcome with 95% confidence: reject h_0\n",
      "    two-sided p-value:           <1e-66\n",
      "\n",
      "Details:\n",
      "    number of observations:      497\n",
      "    Wilcoxon rank-sum statistic: 116375.0\n",
      "    rank sums:                   [116375.0, 6385.0]\n",
      "    adjustment for ties:         0.0\n",
      "    normal approximation (μ, σ): (54995.0, 3184.01)\n",
      "\n",
      "Fraction Expressing cells\n",
      "WT 8H vs Rad21KO 8H\n",
      "Approximate Wilcoxon signed rank test\n",
      "-------------------------------------\n",
      "Population details:\n",
      "    parameter of interest:   Location parameter (pseudomedian)\n",
      "    value under h_0:         0\n",
      "    point estimate:          0.0384241\n",
      "    95% confidence interval: (0.049, 0.0738)\n",
      "\n",
      "Test summary:\n",
      "    outcome with 95% confidence: reject h_0\n",
      "    two-sided p-value:           <1e-26\n",
      "\n",
      "Details:\n",
      "    number of observations:      497\n",
      "    Wilcoxon rank-sum statistic: 94333.5\n",
      "    rank sums:                   [94333.5, 26452.5]\n",
      "    adjustment for ties:         138.0\n",
      "    normal approximation (μ, σ): (33940.5, 3145.53)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3.842026142569548e-27"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "println(\"ln CPM plus 1 in expressing cells\")\n",
    "println(\"WT 8H vs Rad21KO 8H\")\n",
    "using HypothesisTests\n",
    "t = HypothesisTests.SignedRankTest([ii for ii in sceBhatt[!,\"WT_8H__mu\"]], [ii for ii in sceBhatt[!,\"RAD21_8H__mu\"]])\n",
    "println(t)\n",
    "p3 = pvalue(t)\n",
    "println(\"Fraction Expressing cells\")\n",
    "\n",
    "println(\"WT 8H vs Rad21KO 8H\")\n",
    "using HypothesisTests\n",
    "t = HypothesisTests.SignedRankTest([ii for ii in sceBhatt[!,\"WT_8H__alpha\"]], [ii for ii in sceBhatt[!,\"RAD21_8H__alpha\"]])\n",
    "println(t)\n",
    "p3a = pvalue(t)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: both Pandas and PyPlot export \"plot\"; uses of it in module Main must be qualified\n"
     ]
    },
    {
     "ename": "LoadError",
     "evalue": "UndefVarError: plot not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: plot not defined",
      "",
      "Stacktrace:",
      " [1] top-level scope",
      "   @ ./In[11]:19",
      " [2] eval",
      "   @ ./boot.jl:360 [inlined]",
      " [3] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)",
      "   @ Base ./loading.jl:1094"
     ]
    }
   ],
   "source": [
    "figure(figsize = (7, 6))\n",
    "\n",
    "\n",
    "subplot(1,2,2)\n",
    "y = \"ln(CPM + 1) \\n cells with transcripts detected\"\n",
    "pd = Pandas.DataFrame(sort!(sceBhatt, :Class))\n",
    "\n",
    "pd = Pandas.melt(pd, value_vars  = [\"WT_UT__mu\", \"RAD21_UT__mu\" ,\"WT_2H__mu\", \"RAD21_2H__mu\", \"WT_8H__mu\", \"RAD21_8H__mu\"], value_name = y, id_vars = [\"GeneID\",\"Class\"] )\n",
    "pd[\"Time after LPS (h)\"] = [replace(replace(replace(split(ii, \"_\")[2], \"UT\" => 0), \"2H\" => 2), \"8H\" => 8) for ii in pd[\"variable\"]]\n",
    "pd[\"Genotype\"] = [replace(split(ii, \"_\")[1], \"RAD21\" =>\"Rad21KO\") for ii in pd[\"variable\"]]\n",
    "\n",
    "Seaborn.boxplot(data = pd, y = y,x = \"Time after LPS (h)\", hue = \"Genotype\",showfliers = false,  palette = [\"darkgray\", \"red\"])\n",
    "pretty_axes2()\n",
    "\n",
    "hs = [8.2, 8.7, 9.0]\n",
    "ps = adjust([p1,p2,p3], Bonferroni())\n",
    "\n",
    "for ii in 1:length(hs)\n",
    "plot([-0.25+ii-1, 0.25+ii-1], [hs[ii], hs[ii]], lw = 0.75, c= \"black\")\n",
    "annotate(\"P = \"*string(round(ps[ii], sigdigits = 3)),xy = [ii-1, hs[ii]+0.1], va = \"center\", ha = \"center\")\n",
    "end\n",
    "squareplot()\n",
    "\n",
    "ax = gca()\n",
    "for line in ax.get_lines()\n",
    "    line.set_color(\"black\")\n",
    "end\n",
    "subplot(1,2,1)\n",
    "\n",
    "\n",
    "\n",
    "y = \"Fraction of cells \\n with transcripts detected\"\n",
    "pd = Pandas.DataFrame(sort!(sceBhatt, :Class))\n",
    "\n",
    "pd = Pandas.melt(pd, value_vars  = [\"WT_UT__alpha\", \"RAD21_UT__alpha\" ,\"WT_2H__alpha\", \"RAD21_2H__alpha\", \"WT_8H__alpha\", \"RAD21_8H__alpha\"], value_name = y, id_vars = [\"GeneID\",\"Class\"] )\n",
    "pd[\"Time after LPS (h)\"] = [replace(replace(replace(split(ii, \"_\")[2], \"UT\" => 0), \"2H\" => 2), \"8H\" => 8) for ii in pd[\"variable\"]]\n",
    "pd[\"Genotype\"] = [replace(split(ii, \"_\")[1], \"RAD21\" =>\"Rad21KO\") for ii in pd[\"variable\"]]\n",
    "\n",
    "Seaborn.boxplot(data = pd, y = y,x = \"Time after LPS (h)\", hue = \"Genotype\",showfliers = false, palette = [\"darkgray\", \"red\"])\n",
    "pretty_axes2()\n",
    "\n",
    "hs = [0.6, 1.05, 1.05]\n",
    "ps = adjust([p1a,p2a,p3a], Bonferroni())\n",
    "\n",
    "for ii in 1:length(hs)\n",
    "plot([-0.25+ii-1, 0.25+ii-1], [hs[ii], hs[ii]], lw = 0.75, c= \"black\")\n",
    "annotate(\"P = \"*string(round(ps[ii], sigdigits = 2)),xy = [ii-1, hs[ii]+0.05], va = \"center\", ha = \"center\")\n",
    "end\n",
    "\n",
    "legend_removal()\n",
    "\n",
    "ylim(-0.05, 1.1)\n",
    "squareplot()\n",
    "ax = gca()\n",
    "for line in ax.get_lines()\n",
    "    line.set_color(\"black\")\n",
    "end\n",
    "plt.tight_layout()\n",
    "savefigwithtext(\"figures/scRNAseq_mu_alpha_bhattgenes_WTRad21KO.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "PyError ($(Expr(:escape, :(ccall(#= /Users/irr15/.julia/packages/PyCall/BD546/src/pyeval.jl:38 =# @pysym(:PyEval_EvalCode), PyPtr, (PyPtr, PyPtr, PyPtr), o, globals, locals))))) <class 'FileNotFoundError'>\nFileNotFoundError(2, 'No such file or directory')\n  File \"/Users/irr15/.julia/packages/PyCall/BD546/src/pyeval.jl\", line 8, in <module>\n    get!(_namespaces, m) do\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/pyplot.py\", line 859, in savefig\n    res = fig.savefig(*args, **kwargs)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/figure.py\", line 2311, in savefig\n    self.canvas.print_figure(fname, **kwargs)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backend_bases.py\", line 2210, in print_figure\n    result = print_method(\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backend_bases.py\", line 1639, in wrapper\n    return func(*args, **kwargs)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backends/backend_pdf.py\", line 2586, in print_pdf\n    file = PdfFile(filename, metadata=metadata)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backends/backend_pdf.py\", line 562, in __init__\n    fh, opened = cbook.to_filehandle(filename, \"wb\", return_opened=True)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/cbook/__init__.py\", line 408, in to_filehandle\n    fh = open(fname, flag, encoding=encoding)\n",
     "output_type": "error",
     "traceback": [
      "PyError ($(Expr(:escape, :(ccall(#= /Users/irr15/.julia/packages/PyCall/BD546/src/pyeval.jl:38 =# @pysym(:PyEval_EvalCode), PyPtr, (PyPtr, PyPtr, PyPtr), o, globals, locals))))) <class 'FileNotFoundError'>\nFileNotFoundError(2, 'No such file or directory')\n  File \"/Users/irr15/.julia/packages/PyCall/BD546/src/pyeval.jl\", line 8, in <module>\n    get!(_namespaces, m) do\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/pyplot.py\", line 859, in savefig\n    res = fig.savefig(*args, **kwargs)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/figure.py\", line 2311, in savefig\n    self.canvas.print_figure(fname, **kwargs)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backend_bases.py\", line 2210, in print_figure\n    result = print_method(\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backend_bases.py\", line 1639, in wrapper\n    return func(*args, **kwargs)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backends/backend_pdf.py\", line 2586, in print_pdf\n    file = PdfFile(filename, metadata=metadata)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/backends/backend_pdf.py\", line 562, in __init__\n    fh, opened = cbook.to_filehandle(filename, \"wb\", return_opened=True)\n  File \"/Users/irr15/.julia/conda/3/lib/python3.8/site-packages/matplotlib/cbook/__init__.py\", line 408, in to_filehandle\n    fh = open(fname, flag, encoding=encoding)\n",
      "",
      "Stacktrace:",
      "  [1] pyerr_check",
      "    @ ~/.julia/packages/PyCall/BD546/src/exception.jl:62 [inlined]",
      "  [2] pyerr_check",
      "    @ ~/.julia/packages/PyCall/BD546/src/exception.jl:66 [inlined]",
      "  [3] _handle_error(msg::String)",
      "    @ PyCall ~/.julia/packages/PyCall/BD546/src/exception.jl:83",
      "  [4] macro expansion",
      "    @ ~/.julia/packages/PyCall/BD546/src/exception.jl:97 [inlined]",
      "  [5] #117",
      "    @ ~/.julia/packages/PyCall/BD546/src/pyeval.jl:38 [inlined]",
      "  [6] disable_sigint",
      "    @ ./c.jl:458 [inlined]",
      "  [7] pyeval_(s::String, globals::PyDict{String, PyObject, true}, locals::PyDict{String, PyObject, true}, input_type::Int64, fname::String)",
      "    @ PyCall ~/.julia/packages/PyCall/BD546/src/pyeval.jl:37",
      "  [8] macro expansion",
      "    @ ~/.julia/packages/PyCall/BD546/src/pyeval.jl:230 [inlined]",
      "  [9] savefigwithtext(figname::String)",
      "    @ PrettyPlotting ~/Dropbox/Code/PrettyPlotting/src/PrettyPlotting.jl:94",
      " [10] top-level scope",
      "    @ In[12]:32",
      " [11] eval",
      "    @ ./boot.jl:360 [inlined]",
      " [12] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)",
      "    @ Base ./loading.jl:1094"
     ]
    }
   ],
   "source": [
    "col_colors_dict = Dict(\n",
    "    \"WT_UT\" => \"#f0f0f0\",\n",
    "    \"WT_2H\" => \"#bdbdbd\",\n",
    "    \"WT_8H\" => \"#636363\",\n",
    "    \"RAD21_UT\" => \"#fee0d2\",\n",
    "    \"RAD21_2H\" => \"#fc9272\",\n",
    "    \"RAD21_8H\" => \"#de2d26\",\n",
    "    \"RAD21\" => \"red\",\n",
    "    \"WT\" => \"darkgrey\",\n",
    "    0 => \"#deebf7\",\n",
    "    2 => \"#9ecae1\",\n",
    "    8 => \"#3182bd\",\n",
    "    ) \n",
    "\n",
    "\n",
    "sce.colData[!,\"LPS\"] = [replace(replace(replace(split(ii, \"_\")[2], \"UT\" => 0), \"2H\" => 2), \"8H\" => 8) for ii in sce.colData[!,\"Sample\"]]\n",
    "\n",
    "\n",
    "sce = sort_cells!(sce, cols = [ :LPS, :Genotype,:BhattGenesScore], rev = [false, true, false])\n",
    "subsce = select_these_genes(bhattgenes[!,:GeneID], sce)\n",
    "\n",
    "g = Seaborn.clustermap(subsce.assays[\"lnCPMplus1\"], figsize = (7, 4),yticklabels=false,xticklabels=false, col_cluster = false, col_colors = [col_colors_dict[ii] for ii in subsce.colData[!,:Genotype]], cmap=\"coolwarm\")\n",
    "\n",
    "ax = g.ax_heatmap\n",
    "ax.set_ylabel(\"LPS inducible genes\")\n",
    "ax.set_xlabel(\"Cells\")\n",
    "ax.tick_params(axis=\"both\", which=\"both\", length=0)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "savefigwithtext(\"figures/scRNAseq_heatmap_bhattgenes_WTRad21KO.pdf\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.0",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
